vmap fallback kernel

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42324 vmap: temporarily disable support for (most) random functions
* #42248 Implement batching rules for some view ops
* **#41943 vmap fallback kernel**

If an operator doesn't have a batching rule implemented then we fallback
to this implementation. The fallback only works on out-of-place operators
that return only tensors with new memory. (e.g., no in-place operators,
no view operations).

The fallback effectively takes all of the BatchedTensors in `stack`,
slices them, and runs `op` on all of the corresponding slices to produce slices
of the outputs. The output slices then get `torch.stack`ed to create the
final returns.

The performance of the fallback is not very good because it introduces
an extra copy from stacking the sliced outputs. Because of this, we prefer
to write batching rules for operators whenever possible.

In the future, I'd like to disable the fallback kernel for random
functions until we have a better random story for vmap. I will probably
add a blocklist of operators to support that.

Test Plan:
- `pytest test/test_vmap.py -v`

Differential Revision: [D22764103](https://our.internmc.facebook.com/intern/diff/D22764103)

zou3519 committed:
vmap fallback kernel

If an operator doesn't have a batching rule implemented then we fallback
to this implementation. The fallback only works on out-of-place operators
that return only tensors with new memory. (e.g., no in-place operators,
no view operations).

The fallback effectively takes all of the BatchedTensors in `stack`,
slices them, and runs `op` on all of the corresponding slices to produce slices
of the outputs. The output slices then get `torch.stack`ed to create the
final returns.

The performance of the fallback is not very good because it introduces
an extra copy from stacking the sliced outputs. Because of this, we prefer
to write batching rules for operators whenever possible.

In the future, I'd like to disable the fallback kernel for random
functions until we have a better random story for vmap. I will probably
add a blocklist of operators to support that.

Test Plan:
- `pytest test/test_vmap.py -v`

[ghstack-poisoned]

zou3519 committed:
Update on "vmap fallback kernel"


If an operator doesn't have a batching rule implemented then we fallback
to this implementation. The fallback only works on out-of-place operators
that return only tensors with new memory. (e.g., no in-place operators,
no view operations).

The fallback effectively takes all of the BatchedTensors in `stack`,
slices them, and runs `op` on all of the corresponding slices to produce slices
of the outputs. The output slices then get `torch.stack`ed to create the
final returns.

The performance of the fallback is not very good because it introduces
an extra copy from stacking the sliced outputs. Because of this, we prefer
to write batching rules for operators whenever possible.

In the future, I'd like to disable the fallback kernel for random
functions until we have a better random story for vmap. I will probably
add a blocklist of operators to support that.

Test Plan:
- `pytest test/test_vmap.py -v`

[ghstack-poisoned]

zou3519 committed:
Update on "vmap fallback kernel"


If an operator doesn't have a batching rule implemented then we fallback
to this implementation. The fallback only works on out-of-place operators
that return only tensors with new memory. (e.g., no in-place operators,
no view operations).

The fallback effectively takes all of the BatchedTensors in `stack`,
slices them, and runs `op` on all of the corresponding slices to produce slices
of the outputs. The output slices then get `torch.stack`ed to create the
final returns.

The performance of the fallback is not very good because it introduces
an extra copy from stacking the sliced outputs. Because of this, we prefer
to write batching rules for operators whenever possible.

In the future, I'd like to disable the fallback kernel for random
functions until we have a better random story for vmap. I will probably
add a blocklist of operators to support that.

Test Plan:
- `pytest test/test_vmap.py -v`

[ghstack-poisoned]

zou3519 committed:
Update on "vmap fallback kernel"


If an operator doesn't have a batching rule implemented then we fallback
to this implementation. The fallback only works on out-of-place operators
that return only tensors with new memory. (e.g., no in-place operators,
no view operations).

The fallback effectively takes all of the BatchedTensors in `stack`,
slices them, and runs `op` on all of the corresponding slices to produce slices
of the outputs. The output slices then get `torch.stack`ed to create the
final returns.

The performance of the fallback is not very good because it introduces
an extra copy from stacking the sliced outputs. Because of this, we prefer
to write batching rules for operators whenever possible.

In the future, I'd like to disable the fallback kernel for random
functions until we have a better random story for vmap. I will probably
add a blocklist of operators to support that.

Test Plan:
- `pytest test/test_vmap.py -v`

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/BatchedFallback.cpp:
```diff
@@ -0,0 +1,127 @@
+  std::vector<Tensor> batched_tensor_inputs;
+  VmapDimVector batched_tensor_inputs_position;
+  for (int64_t idx = 0; idx < stack->size(); ++idx) {
+    const auto& ivalue = (*stack)[idx];
```
Is this morally OK? `torch::jit::Stack` is typedef'ed to be `vector<ivalue>`: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/stack.h#L11 and I am using it as a vector throughout this function.

There are a lot of convenience functions for `torch::jit::Stack` (https://github.com/pytorch/pytorch/blob/b85df3709a9e92f88d2e86b4860571010bc06e62/aten/src/ATen/core/stack.h#L29-L49) but they're harder to use in this use case than the `std::vector` API.

zou3519 replied:
Answered: morally incorrect.

zou3519 commented on aten/src/ATen/BatchedFallback.cpp:
```diff
@@ -0,0 +1,127 @@
+    // ivalue is a BatchedTensor
+    const auto& physical_view_for_ivalue = *input_physical_views_iter;
+    for (int64_t linear_idx = 0; linear_idx < num_batches; ++linear_idx) {
+      auto index = computeIndex(linear_idx, batch_sizes);
```
Instead of computing the index every time, I could modify the index in-place and add 1 to it. Is this worth it? In terms of performance it looks like it is more efficient

ezyang replied:
I don't think I care too much, either way.

zou3519 replied:
I'll stick with the current approach (re-compute the index vector) for now and optimize later if necessary

zou3519 commented on aten/src/ATen/BatchedFallback.cpp:
```diff
@@ -0,0 +1,127 @@
+    TORCH_INTERNAL_ASSERT(stack.size() == 1)
+    output_shards.push_back(stack[0].toTensor());
+  }
+  auto flat_output = at::stack(output_shards);
```
The `stack` means that the fallback kernel won't be as performant as a user deciding to use a for-loop rather than vmap. However, the long-term state for vmap is to not rely on the fallback (and instead have batching rules defined on the operations) so I am not too worried about the performance.

## Problem

Consider the following model:
```
def model(example):
    example = F.conv(example, ...)
    example = F.relu(example, ...)
    example = F.linear(example, ...)
    return tensor
```
If the user wanted to push multiple examples via a Python for-loop, they’d do the following:
```
examples = torch.randn(N, ...)
results = []
for example in examples.unbind(0):
    results.append(model(example))
output = torch.stack(results)
```
Now, let’s assume that no batching rules were implemented for F.conv, F.relu, and F.linear. The vmap “for-loop” fallback would do the following, for each operation:

* unbind the examples along the batch dimension
* run the operator for each example
* stack the results

In other words, semantically, vmap(model)(examples) is equivalent to:
```
examples = torch.stack([F.conv(example, ...) for example in examples])
examples = torch.stack([F.relu(example, ...) for example in examples])
examples = torch.stack([F.linear(example, ...) for example in examples])
```
All of the torch.stack make the vmap “for-loop” fallback slower than the python for-loop, which is not very good for vmap.

ezyang replied:
The obvious fix is to not stack: instead, use `out` to write samples directly into a preallocated output. But to preallocate the output, you need a way of computing what the size of the computation will be on one output, which is not so easy to do right now. One hacky way to do it would be to run on one example, then resize the output so it's large enough to fit the rest of the examples, and then run the loop. It'll be faster than stack. Then if we have working meta functions, you can use those to compute what the output size should be after the fact.

zou3519 replied:
We can't actually use the `out=` variant of the function because they doesn't support autograd. Because vmap occurs above autograd in the current design, vmap must desugar into operations that are compatible with autograd.

ezyang replied:
RIP. cc @albanD

albanD replied:
Note that this will depend on the overall model: a short model (few ops) won't see much difference while a deeper model will be more impacted.
Also I guess that the output size vs complexity of the function itself will have a big impact here.

But in general, for functions with inplace versions (`foo_()`), we could do this and keep autograd. But we will need a list of such functions and register a different fallback kernel for these.
For other functions, I can't think of a better way to do it :/

ezyang replied:
@albanD what is being asked for here is a little different. Imagine:

```
x1 = torch.randn(2, requires_grad=True)
x2 = torch.randn(2, requires_grad=True)
x3 = torch.randn(2, requires_grad=True)
y = torch.empty(3, 2)
x1.add(w, out=y[0])
x2.add(w, out=y[1])
x3.add(w, out=y[2])
y.backward()
```

It would be cool if this worked.

albanD replied:
Yes it would :D 
As mentioned in the `out=` autograd proposal [here](https://github.com/pytorch/pytorch/issues/29378#issuecomment-571805569), that would work just fine.
Unfortunately, detecting these valid cases and implementing this overall is not going to be done any time soon I think.

The version for the `foo_()` variants would work with the current code base though (even though it won't work with as many ops).

zou3519 committed:
Update on "vmap fallback kernel"


If an operator doesn't have a batching rule implemented then we fallback
to this implementation. The fallback only works on out-of-place operators
that return only tensors with new memory. (e.g., no in-place operators,
no view operations).

The fallback effectively takes all of the BatchedTensors in `stack`,
slices them, and runs `op` on all of the corresponding slices to produce slices
of the outputs. The output slices then get `torch.stack`ed to create the
final returns.

The performance of the fallback is not very good because it introduces
an extra copy from stacking the sliced outputs. Because of this, we prefer
to write batching rules for operators whenever possible.

In the future, I'd like to disable the fallback kernel for random
functions until we have a better random story for vmap. I will probably
add a blocklist of operators to support that.

Test Plan:
- `pytest test/test_vmap.py -v`

[ghstack-poisoned]

ezyang commented on aten/src/ATen/BatchedFallback.cpp:
```diff
@@ -0,0 +1,142 @@
+      1,
+      std::multiplies<int64_t>());
+
+  // Populate `num_batches` number of torch::jit::Stack, one for each computation.
```
A more idiomatic way to do this is to have a single stack, and then push/pop elements off of it each time you do the loop. I don't know how much simpler precomputing is, but if you maintain only a single stack the amount of memory you use and the number of dynamic allocations you do is reduced. (By the way, this is why it's a stack in the first place: so we can reuse a contiguous space of memory to do calls.

ezyang replied:
I'm not sure if this is more complicated to do or not, but I don't feel it should be that much more complicated. Now you are iterating over batch, and then for each batch, you construct the correct arguments for running the batch and then make the call. In fact, the current version of the code feels "transposed" to me.

zou3519 replied:
Hmm, let me think about that

zou3519 replied:
Yes, you're right -- the logic in this PR is "transposed". Transposing it again wasn't very difficult

ezyang commented on aten/src/ATen/BatchedFallback.cpp:
```diff
@@ -0,0 +1,142 @@
+      batched_tensor_inputs);
+
+  // Compute the total number of batches
+  auto num_batch_dims = input_physical_views.front().numBatchDims();
```
Hmm, it's a little bit of a smell that we arbitrarily read out the first view for the number of batch dims, when in fact all of the views are guaranteed to have the same value (right?) Not sure if this is actionable though.

zou3519 replied:
Yes, all of the views have the same `numBatchDims()`. I'll think about if we can fix this somehow but my initial thought is that we might have to change the VmapTransforms API

ezyang commented on test/test_vmap.py:
```diff
@@ -433,5 +445,62 @@ def foo(x):
+    def test_fallback_masked_fill(self):
+        # NB: One day we will implement a batching rule for index_add
+        # If/when we do, this test should be replaced to test the fallback
+        # path on another operator to avoid bitrot.
```
It would be better if you could detect if this situation occurred :) (Maybe you could use one of the apis used in `test_dispatch.py`, or a new API, to test that `index_add` doesn't have an explicit VMap key installed. Not sure it's worth the faffing about.)

zou3519 replied:
We can detect it actually -- the fallback emits a warning message and we test for the warning message in `test_fallback_masked_fill`. If someone adds a batching rule for `masked_fill`, then the test will fail and the developer will read this comment :)

zou3519 committed:
Update on "vmap fallback kernel"


If an operator doesn't have a batching rule implemented then we fallback
to this implementation. The fallback only works on out-of-place operators
that return only tensors with new memory. (e.g., no in-place operators,
no view operations).

The fallback effectively takes all of the BatchedTensors in `stack`,
slices them, and runs `op` on all of the corresponding slices to produce slices
of the outputs. The output slices then get `torch.stack`ed to create the
final returns.

The performance of the fallback is not very good because it introduces
an extra copy from stacking the sliced outputs. Because of this, we prefer
to write batching rules for operators whenever possible.

In the future, I'd like to disable the fallback kernel for random
functions until we have a better random story for vmap. I will probably
add a blocklist of operators to support that.

Test Plan:
- `pytest test/test_vmap.py -v`

Differential Revision: [D22764103](https://our.internmc.facebook.com/intern/diff/D22764103)

[ghstack-poisoned]

zou3519 committed:
Update on "vmap fallback kernel"


If an operator doesn't have a batching rule implemented then we fallback
to this implementation. The fallback only works on out-of-place operators
that return only tensors with new memory. (e.g., no in-place operators,
no view operations).

The fallback effectively takes all of the BatchedTensors in `stack`,
slices them, and runs `op` on all of the corresponding slices to produce slices
of the outputs. The output slices then get `torch.stack`ed to create the
final returns.

The performance of the fallback is not very good because it introduces
an extra copy from stacking the sliced outputs. Because of this, we prefer
to write batching rules for operators whenever possible.

In the future, I'd like to disable the fallback kernel for random
functions until we have a better random story for vmap. I will probably
add a blocklist of operators to support that.

Test Plan:
- `pytest test/test_vmap.py -v`

Differential Revision: [D22764103](https://our.internmc.facebook.com/intern/diff/D22764103)

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/BatchedFallback.cpp:
```diff
@@ -0,0 +1,142 @@
+  // For each BatchedTensor, also record what position of `stack` they came from.
+  std::vector<Tensor> batched_tensor_inputs;
+  VmapDimVector batched_tensor_inputs_position;
+  for (int64_t idx = 0; idx < stack->size(); ++idx) {
```
fixed using a more idiomatic `torch::jit::last(stack, n)`

zou3519 commented on aten/src/ATen/BatchedFallback.cpp:
```diff
@@ -0,0 +1,137 @@
+  // Strategy: For each batch, we are going to push slices (where applicable)
+  // of the arguments onto `working_stack`, call `op`, and store the result in
+  // `output_shards`.
+  torch::jit::Stack working_stack;
```
@ezyang I wanted to take this a step further and reuse the input stack. However, replacing this line with `auto& working_stack = *stack` causes some difficult-to-debug error messages like "RuntimeError: isTensor() INTERNAL ASSERT FAILED at "../aten/src/ATen/core/ivalue_inl.h":117, please report a bug to PyTorch. Expected Tensor but got None", "RuntimeError: IValue is not a Scalar". Perhaps there is existing code somewhere that assumes the stack must not contain additional arguments unrelated to the current operation?

ezyang replied:
Can you get a full backtrace (and state of the stack) from gdb here? This sounds unexpected.

zou3519 replied:
I figured out the problem, it was a bug in my code. I've changed the code to just use the input stack.

If you're interested in the bug: somewhere above the code uses `torch::jit::last` to get an `ArrayRef<IValue>` of the inputs on the stack. But it turns out that when you modify the stack by pushing elements to it, the original `ArrayRef<IValue>` becomes garbage because the underlying `std::vector` that backs the stack is allowed to reallocate memory (and it does). My code was reading garbage values and that is how the errors happened.

zou3519 commented:
I updated the strategy to do stack push-pop

zou3519 committed:
Update on "vmap fallback kernel"


If an operator doesn't have a batching rule implemented then we fallback
to this implementation. The fallback only works on out-of-place operators
that return only tensors with new memory. (e.g., no in-place operators,
no view operations).

The fallback effectively takes all of the BatchedTensors in `stack`,
slices them, and runs `op` on all of the corresponding slices to produce slices
of the outputs. The output slices then get `torch.stack`ed to create the
final returns.

The performance of the fallback is not very good because it introduces
an extra copy from stacking the sliced outputs. Because of this, we prefer
to write batching rules for operators whenever possible.

In the future, I'd like to disable the fallback kernel for random
functions until we have a better random story for vmap. I will probably
add a blocklist of operators to support that.

Test Plan:
- `pytest test/test_vmap.py -v`

Differential Revision: [D22764103](https://our.internmc.facebook.com/intern/diff/D22764103)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/BatchedFallback.cpp:
```diff
@@ -0,0 +1,154 @@
+
+    op.callBoxed(stack);
+
+    // We assume there is a single tensor return
```
Hmmmmmmmm

ezyang replied:
This is probably not true in general

zou3519 replied:
Yeah, that'll get fixed in the future :) (probably one of the next three diffs)

zou3519 replied:
I tried implementing support for multiple returns but I ran into https://github.com/pytorch/pytorch/issues/42378 (there's a minimal repro in the example). Should an operator with multiple returns push all of them onto the stack?

ezyang commented on aten/src/ATen/BatchedFallback.cpp:
```diff
@@ -0,0 +1,154 @@
+      // At this point, the stack should look like the following:
+      // [..., A, B, C, D, A', B'] (top of stack here)
+      // To get at C (argument number 2), we have to peek `num_arguments`
+      // into the stack from the top
```
TBH, you should just remember where the top of the stack was when you entered and then just index off of that

ezyang replied:
(You don't want a from top of stack accessor here)

zou3519 replied:
Are you suggesting to store a pointer to the top of the stack? The problem with a pointer is that the memory location can change due to reallocations as you add new elements onto the stack due to how `std::vector` works

zou3519 replied:
(answered offline): Ed's suggestion was to store an index

zou3519 committed:
Update on "vmap fallback kernel"


If an operator doesn't have a batching rule implemented then we fallback
to this implementation. The fallback only works on out-of-place operators
that return only tensors with new memory. (e.g., no in-place operators,
no view operations).

The fallback effectively takes all of the BatchedTensors in `stack`,
slices them, and runs `op` on all of the corresponding slices to produce slices
of the outputs. The output slices then get `torch.stack`ed to create the
final returns.

The performance of the fallback is not very good because it introduces
an extra copy from stacking the sliced outputs. Because of this, we prefer
to write batching rules for operators whenever possible.

In the future, I'd like to disable the fallback kernel for random
functions until we have a better random story for vmap. I will probably
add a blocklist of operators to support that.

Test Plan:
- `pytest test/test_vmap.py -v`

Differential Revision: [D22764103](https://our.internmc.facebook.com/intern/diff/D22764103)

[ghstack-poisoned]

zou3519 committed:
Update on "vmap fallback kernel"


If an operator doesn't have a batching rule implemented then we fallback
to this implementation. The fallback only works on out-of-place operators
that return only tensors with new memory. (e.g., no in-place operators,
no view operations).

The fallback effectively takes all of the BatchedTensors in `stack`,
slices them, and runs `op` on all of the corresponding slices to produce slices
of the outputs. The output slices then get `torch.stack`ed to create the
final returns.

The performance of the fallback is not very good because it introduces
an extra copy from stacking the sliced outputs. Because of this, we prefer
to write batching rules for operators whenever possible.

In the future, I'd like to disable the fallback kernel for random
functions until we have a better random story for vmap. I will probably
add a blocklist of operators to support that.

Test Plan:
- `pytest test/test_vmap.py -v`

Differential Revision: [D22764103](https://our.internmc.facebook.com/intern/diff/D22764103)

[ghstack-poisoned]

zou3519 committed:
Update on "vmap fallback kernel"


If an operator doesn't have a batching rule implemented then we fallback
to this implementation. The fallback only works on out-of-place operators
that return only tensors with new memory. (e.g., no in-place operators,
no view operations).

The fallback effectively takes all of the BatchedTensors in `stack`,
slices them, and runs `op` on all of the corresponding slices to produce slices
of the outputs. The output slices then get `torch.stack`ed to create the
final returns.

The performance of the fallback is not very good because it introduces
an extra copy from stacking the sliced outputs. Because of this, we prefer
to write batching rules for operators whenever possible.

In the future, I'd like to disable the fallback kernel for random
functions until we have a better random story for vmap. I will probably
add a blocklist of operators to support that.

Test Plan:
- `pytest test/test_vmap.py -v`

Differential Revision: [D22764103](https://our.internmc.facebook.com/intern/diff/D22764103)

[ghstack-poisoned]
